flaskr.py

```diff
@@ -34,50 +35,105 @@ def init_db():
-	entries = cur.fetchall()
-	return render_template('show_entries.html', entries=entries)
+    db=get_db()
+    cur = db.execute('select id, title, text from entries order by id desc')
+    entries = cur.fetchall()
+    return render_template('show_entries.html', entries=entries)
 
 @app.route('/add', methods=['POST'])
 def add_entry():
-	if not session.get('logged_in'):
-		abort(401)
-	db = get_db()
-	db.execute('insert into entries (title, text) values (?, ?)',
-		[request.form['title'], request.form['text']])
-	db.commit()
-	flash('New entry was successfuly posted')
-	return redirect(url_for('show_entries'))
+    if not session.get('logged_in'):
+        abort(401)
+    db = get_db()
+    db.execute('insert into entries (title, text) values (?, ?)',
+        [request.form['title'], request.form['text']])
+    db.commit()
+    flash('New entry was successfuly posted')
+    return redirect(url_for('show_entries'))
 
 @app.route('/login', methods=['GET','POST'])
 def login():
-	error=None
-	if request.method == 'POST':
-		if request.form['username'] != app.config['USERNAME']:
-			error = 'Invalid username'
-		elif request.form['password'] != app.config['PASSWORD']:
-			error = 'Invalid password'
-		else:
-			session['logged_in'] = True
-			flash('You were logged in')
-			return redirect(url_for('show_entries'))
-	return render_template('login.html', error=error)			
+    error=None
+    if request.method == 'POST':
+        if request.form['username'] != app.config['USERNAME']:
+            error = 'Invalid username'
+        elif request.form['password'] != app.config['PASSWORD']:
+            error = 'Invalid password'
+        else:
+            session['logged_in'] = True
+            flash('You were logged in')
+            return redirect(url_for('show_entries'))
+    return render_template('login.html', error=error)           
 
 @app.route('/logout')
 def logout():
-	session.pop('logged_in', None)
-	flash('You were logged out')
-	return redirect(url_for('show_entries'))
\ No newline at end of file
+    session.pop('logged_in', None)
+    flash('You were logged out')
+    return redirect(url_for('show_entries'))
+
+
+@app.route('/entry=<entry_id>&move=<direction>', methods=['GET'])
+def move_entry(entry_id, direction):
+    """Move entry up or down."""
+    db = get_db()
+    cur = db.execute('select id from entries order by id desc')
+    data = []
+    for i, row in enumerate(cur):
+        data.append([i, row[0]])
+    print(data)
+    if fnmatch.fnmatch(direction, 'up'):
+        if data[0][1] == int(entry_id):
+            flash('The entry is already on the top')
+            return redirect(url_for('show_entries'))
+        else:
+            for i, row in enumerate(data):
+                if row[1] == int(entry_id):
+                        current_row = data[i]
+                        previous_row = data[i - 1]
+                        db.execute('''UPDATE entries
```

Comment:
I figured out almost everything that you asked for, except really generic algorithm, it seems that I'm pretty close to it, but not sure if I changed the code correctly, please check the last commit before I continue working on this.